src/hotspot/share/opto/library_call.cpp

```diff
@@ -5915,8 +5915,16 @@ bool LibraryCallKit::inline_arraycopy() {
     record_for_igvn(slow_region);
 
     // (1) src and dest are arrays.
+    // Keep track of the information that src/dest are arrays to prevent below array specific accesses from floating above.
     generate_non_array_guard(load_object_klass(src), slow_region);
+    if (!stopped()) {
+      src = _gvn.transform(new CheckCastPPNode(control(), src, TypeAryPtr::BOTTOM));
```

Comment:
Why is this a `CheckCastPP` and not a `CastPP`? My understanding is that a `CheckCastPP` is used when we force changing the type of a node (e.g a raw pointer of `Allocate` into a typed pointer), so we do not join the type of the input with that of the output.

Reply:
Good point, I changed that.